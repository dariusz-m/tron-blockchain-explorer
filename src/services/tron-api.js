import HttpClient from "@tronprotocol/wallet-api/src/client/http";

export const TronBlockchainClient = new HttpClient();
