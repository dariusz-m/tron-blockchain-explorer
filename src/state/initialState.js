const initialState = {
    router: {"location": null},
    tokens: [],
    selectedBlock: {number: -1, witnessId: -1, hash: "", parentHash: ""}
};

export default initialState;
