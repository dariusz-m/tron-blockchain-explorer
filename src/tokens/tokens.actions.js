export const TokensActionTypes = {
    LOAD_TOKENS: "LOAD_TOKENS",
    TOKENS_HAVE_BEEN_LOADED: "TOKENS_HAVE_BEEN_LOADED",
    TOKENS_HAVE_NOT_BEEN_LOADED: "TOKENS_HAVE_NOT_BEEN_LOADED",
};

export const loadTokens = () => {
    return {
        type: TokensActionTypes.LOAD_TOKENS
    };
};

export const tokensHaveBeenLoaded = (tokens) => {
    return {
        type: TokensActionTypes.TOKENS_HAVE_BEEN_LOADED,
        tokens
    };
};

export const tokensHaveNotBeenLoaded = () => {
    return {
        type: TokensActionTypes.TOKENS_HAVE_NOT_BEEN_LOADED,
    };
};
