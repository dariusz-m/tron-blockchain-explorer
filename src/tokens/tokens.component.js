import React from 'react';
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import {bindActionCreators} from "redux";

import {loadTokens} from './tokens.actions';

export class Tokens extends React.Component {
    componentDidMount() {
        this.props.actions.loadTokens();
    }

    render() {
        return (
            <div>
                <h1>Tokens</h1>
                {this.props.tokens.map((token) => {
                    return (
                        <div key={token.name}>{token.name}</div>
                    );
                })}
            </div>
        );
    }
}

Tokens.propTypes = {
    tokens: PropTypes.arrayOf(
        PropTypes.shape({
            name: PropTypes.string.isRequired,
            ownerAddress: PropTypes.string.isRequired,
            totalSupply: PropTypes.number.isRequired,
            startTime: PropTypes.number.isRequired,
            endTime: PropTypes.number.isRequired,
            description: PropTypes.string.isRequired,
            num: PropTypes.number.isRequired,
            trxNum: PropTypes.number.isRequired,
            price: PropTypes.number.isRequired,
        })
    ).isRequired,
    actions: PropTypes.shape({
        loadTokens: PropTypes.func.isRequired,
    }),
};

function mapStateToProps(state) {
    return {
        tokens: state.tokens
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            loadTokens: loadTokens,
        }, dispatch)
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Tokens);
