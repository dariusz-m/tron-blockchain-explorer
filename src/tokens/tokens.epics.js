import { from, of } from 'rxjs';
import { mergeMap, map, catchError} from 'rxjs/operators';

import {TronBlockchainClient} from '../services/tron-api';
import {TokensActionTypes, tokensHaveBeenLoaded, tokensHaveNotBeenLoaded} from './tokens.actions';

export const loadTokens = action$ =>
    action$.ofType(TokensActionTypes.LOAD_TOKENS)
        .pipe(
            mergeMap(() =>
                from(TronBlockchainClient.getAssetIssueList()).pipe(
                    map((tokens) => tokensHaveBeenLoaded(tokens)),
                    catchError((error) => {
                        console.error(error);
                        return of(tokensHaveNotBeenLoaded());
                    })
                )

            ),
        );

