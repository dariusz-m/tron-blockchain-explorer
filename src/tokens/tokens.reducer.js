import initialState from '../state/initialState';
import {TokensActionTypes} from './tokens.actions';

export default (state = initialState.tokens, action) => {
    switch (action.type) {
        case TokensActionTypes.TOKENS_HAVE_BEEN_LOADED:
            return [...action.tokens];
        default:
            return state;
    }
};
