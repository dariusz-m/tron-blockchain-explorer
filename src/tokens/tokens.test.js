import React from 'react';
import renderer from 'react-test-renderer';

import {Tokens} from "./tokens.component";
import configureStore from "../state";
import {loadTokens} from "./tokens.actions";
import initialState from "../state/initialState";
import {waitForAsynchronousActions} from '../tests.support';

jest.mock('../services/tron-api');
import {TronBlockchainClient} from '../services/tron-api';

describe('Tokens tests', () => {
    let tokens = [];

    beforeEach(() => {
        tokens = [
            {
                name: 'Feefree',
                ownerAddress: '27VvyqLSrvqhRnWSDETwUc9zxF7QWLgKJEG',
                totalSupply: 1000000,
                startTime: 1526242493000,
                endTime: 1534018253000,
                description: 'Free Smartphone with Free monthly plan. Over 90 million qualify!',
                num: 2,
                trxNum: 1000000,
                price: 500000
            },
            {
                name: 'BABA',
                ownerAddress: '27g5kEhMeaSD6vZ4KX5QUCwR6XfYax7xswY',
                totalSupply: 100000,
                startTime: 1526360443000,
                endTime: 1534050043000,
                description: 'Test',
                num: 10,
                trxNum: 1000000,
                price: 100000
            },
        ];
    });

    it('Renders correctly without tokens', () => {
        const actions = {
            loadTokens: () => {},
        };
        const tree = renderer.create(<Tokens actions={actions} tokens={[]}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Renders correctly with tokens', () => {
        const actions = {
            loadTokens: () => {},
        };
        const tree = renderer.create(<Tokens actions={actions} tokens={tokens}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Tokens have been loaded', async () => {
        TronBlockchainClient.getAssetIssueList.mockImplementationOnce(async () => await [...tokens]);
        const store = configureStore();
        const expectedState = {...initialState, tokens: [...tokens]};

        store.dispatch(loadTokens());

        await waitForAsynchronousActions().then(() => expect(store.getState()).toEqual(expectedState));

    });
});

