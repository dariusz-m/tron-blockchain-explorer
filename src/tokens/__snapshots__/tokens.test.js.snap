// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokens tests Renders correctly with tokens 1`] = `
<div>
  <h1>
    Tokens
  </h1>
  <div>
    Feefree
  </div>
  <div>
    BABA
  </div>
</div>
`;

exports[`Tokens tests Renders correctly without tokens 1`] = `
<div>
  <h1>
    Tokens
  </h1>
</div>
`;
